### ISSUE TITLE
Some of the loss functions probably don't work correctly

### ISSUE THREAD

**Issue by iamshnoo** [mlpack/mlpack#2444]
I was verifying some of the loss functions in mlpack before modifying them to introduce reduction functionality in them. Some of them probably aren't implemented correctly or have misleading class names. I am listing these below. The goal of this issue is primarily to discuss any opinions on these. After that, I would like to make the necessary changes to have everything working perfectly.

- [ ] L1 Loss 
- [ ] Cross Entropy Error
- [ ] Mean Bias Error
- [ ] Cosine Embedding Loss
- [ ] Hinge Embedding Loss
- [ ] KL Div Loss
- [ ] Margin Ranking Loss
- [ ] NLL Loss
---

For L1 loss, the forward function doesn't take the absolute difference between input and target. Also, the backward function doesn't return a different output matrix, for different values of the reduction parameter _**mean**_. I have created a demo [Google Colab notebook](https://colab.research.google.com/drive/1EqxCl8jTw7vuWm-0WI_64e64MR6PUzSP?usp=sharing) to demonstrate the changes that need to be done.

---

Cross Entropy Error in mlpack is actually Binary Cross Entropy Loss. Also it works only when target is one-hot encoded and calculates reduction='sum' by default whereas PyTorch calculates reduction='mean' by default. None of these assumptions are written anywhere, inspite of the [discussion](https://github.com/mlpack/mlpack/pull/1070#discussion_r129208108) in #1070 which was merged a long time back. I have verified these claims in a [Google Colab notebook](https://colab.research.google.com/drive/14PxGHVCzNLa9mN4VGluM2p0oQ9OCpCMt?usp=sharing) in case anyone would like to take a quick look at the issue. 

---

Mean Bias Error -> [Google Colab notebook](https://colab.research.google.com/drive/1x8dGBxFl-XtwgPs4EU0-1fnuvLFOp1Xy?usp=sharing)

---

For Cosine Embedding Loss the current mlpack implementation will suffer from 2 warnings. (-Wreorder and -Wunused). 

**-Wreorder** 
bool similarity; double margin; bool takeMean; // original order of definition 
const double margin, const bool similarity, const bool takeMean // constructor order

**-Wunused**
const size_t batchSize = input.n_elem / cols; // in the Backward() method

Also, I have very little idea about how to implement reduction='none' for this function. Will definitely need Kartik's help with it for this function.

---

Dice Loss -> Probably no need to implement reduction for this function (will update later if needed)

---

Earth Mover Distance -> correct implementation (reduction implementation : [Google Colab notebook](https://colab.research.google.com/drive/1xClRmLvuMMcVx5bXatWHJMEgmwR1SRbC?usp=sharing))

---

Hinge Embedding loss -> The current implementation has 2 issues. Primarily, the function is incorrectly implemented and a minor secondary design issue is that it only takes labels that are 0 or 1 instead of labels that are 1 or -1 (as in PyTorch or other frameworks).
[Google Colab notebook](https://colab.research.google.com/drive/1umN57r7_b9wua4_Tz4OArIQthxSnDBIh?usp=sharing)

---

KL Div Loss -> Current implementation has a few issues. I have standardised the implementation to match the PyTorch implementation in this [Google Colab notebook](https://colab.research.google.com/drive/1_57-NMliSNG3xtS3VNy1pXDqHYsHTYHn?usp=sharing) along with correction of those issues.

---

Huber Loss -> Correct implementation. Reduction facilities introduced in this [Google Colab notebook](https://colab.research.google.com/drive/1GxrvluR1aniIlM6d-O0Fq4NubY7uSB3E?usp=sharing)

---

LogCosh Loss -> Correct implementation. Reduction facilities introduced in this [Google Colab notebook](https://colab.research.google.com/drive/1EQhfFZdhYxkcfkScDqcn8lYsWA0EtfzF?usp=sharing)

---

Margin Ranking Loss -> Forward method was incorrect. Fixed it. Backward method is probably incorrect. I don't know how to fix it under current implementation restrictions. See the [Google Colab notebook here](https://colab.research.google.com/drive/1V-g_NmKMuM2EAm79CDwwAXswca1WbE0u?usp=sharing)

---

NLL Loss -> [Google Colab notebook here](https://colab.research.google.com/drive/1_wUfSHE_c4ZsS1dPc5Q4iMhxdXqy_36N?usp=sharing). I am not entirely sure about this. But, what I expect is that for the same input and target matrices, PyTorch and mlpack should be returning the same values, which doesn't happen with the current implementation. (Maybe I am wrong about something?) I have updated the implementation details to match PyTorch outputs and also added all the reduction facilities in the notebook above. 

---

**Comment by mrityunjay-tripathi** [comment]
>For L1 loss, the forward function doesn't take the absolute difference between input and target.

Whoa! Nice catch.
Can you also take a look at `mean_bias_error_impl.hpp`?

**Comment by iamshnoo** [comment]
> Can you also take a look at `mean_bias_error_impl.hpp`?

<img width="594" alt="Screenshot 2020-06-07 at 7 05 19 PM" src="https://user-images.githubusercontent.com/45921510/83970171-d86f9f00-a8f1-11ea-91d7-533b41cf6549.png">

```cpp
// Forward
arma::accu(target - input) / target.n_cols;

// Backward
output.set_size(arma::size(input));
output.fill(-1.0);
```

Seems correct to me.

**Comment by mrityunjay-tripathi** [comment]
Though I haven't gone into detail of it, do you think the `reduction` parameter is missing in `mean_bias_error` or rather it's not required? Also, is it correct to divide by `target.n_cols` in the `Forward` function? It will be all clear once we put this loss function to gradient test and jacobian test. Would you like to do that since you are already involved into it?

**Comment by iamshnoo** [comment]
> do you think the reduction parameter is missing in mean_bias_error

Oh yes, of course. I am currently not looking at that, just checking the loss functions for possible errors. Most of the functions don't currently have reduction parameter. I will be adding those and standardising the implementation across all of the functions in a PR soon.

> is it correct to divide by target.n_cols in the Forward function

Now that you mention it explicitly, it seems it is wrong. 
I tried out what the new implementation would look like in a [Google Colab notebook](). Maybe you would like to take a look at that? (I would also probably do the Jacobian Test for this method soon, but it seems right, because it follows the standard calculations I am following for all the other loss functions, which I have already verified.)

**Comment by mrityunjay-tripathi** [comment]
Thanks for the quick report @iamshnoo. I think every loss function should pass the jacobian test. Let me know what do you think.

**Comment by iamshnoo** [comment]
Yes. That's true. I intend to carry out the Jacobian test for these functions.
However, one thing can be used to advantage to speed up the process. Some of the loss functions already have implementations in PyTorch and since the PyTorch source code already does Jacobian testing, we can just compare the output from PyTorch and mlpack for those functions, instead of re-implementing the wheel!
For the functions which aren't there in PyTorch, I will obviously do the Jacobian tests, and ask you for help if I get stuck somewhere :)

**Comment by mrityunjay-tripathi** [comment]
The problem is who knows PyTorch's implementations don't have bugs? In fact, it happened when I was implementing `poisson_nll_loss` (#2196). I think relying on some other library for tests is risky (but can be used when we don't have other options). One more thing is that armadillo matrices are column-major while PyTorch's tensors are row-major, so that might cause confusion.

**Comment by sreenikSS** [comment]
@mrityunjay-tripathi thanks for pointing it out. @iamshnoo I'd suggest you to fix the errors you have pointed out in this issue. It seems like the Jacobean test would be a big step towards mlpack's self-sufficiency (as we would not need to rely only on pytorch code to check our functions) but I think we should take this up once we have some time in hand because this looks a little time-consuming.

I personally believe that adding Jacobean tests for all the loss functions should be a separate issue in itself (if it doesn't already exists) so that whoever wants to take it up feels free to.

**Comment by kartikdutt18** [comment]
Following @sreenikSS discussions on video meet, It might a nice idea to add PR templates and in that template we could ask for a colab / jupyter notebook link. I generally add the link to reference implementation and the link to tests as shown in #2410, as well as in cosine embedding PR as well. For the loss function that I added, cosine embeddings I think there are only warnings but maybe I missed something. Let me know if I need to create a quick PR to fix it. 
Regards.

**Comment by iamshnoo** [comment]
> For the loss function that I added, cosine embeddings I think there are only warnings but maybe I missed something. 

I have just one doubt regarding [this particular commit](https://github.com/mlpack/mlpack/pull/2209/commits/531358172356636c5e7531e390c28999da70c885) in the cosine embedding loss.

Let me explain with the picture below.
<img  src="https://user-images.githubusercontent.com/45921510/85105951-01186280-b229-11ea-9675-0378510a3178.png">

So, basically if you check that commit I linked above, you will see that the function had 2 explicit inputs previously which were later combined into one using some armadillo magic. 

My point is (let me know if I am completely wrong with this thinking), if a function receives 2 separate input tensors (albeit, combined together as a single matrix), then there will be 2 separate gradient calculations wrt each of the inputs. As from the image above, these will be the matrices g1 and g2. But, currently all the Backward() methods calculate only a single matrix, g. Combining x1 and x2 into a single matrix ```x``` makes sense. But combining g1 and g2 into a single matrix g is probably not correct. This is the reason why the Backward method of Margin Ranking Loss function is also incorrect according to me.

I am not exactly sure about how you have tackled this issue in Cosine Embedding, because when I checked the commit above, the Backward() was still returning a single matrix ```output``` even while the input had 2 explicitly specified tensors. 

If you could take a quick look at the notebook for Margin Ranking Loss that I have shared above, this point should become more clear. (Let me know if I still need to clarify anything after that)

If possible, could you let me know how exactly you have tackled this issue for the Cosine Embedding function or if this issue is not related to your function at all?

Also, it would be super helpful if you could update the Colab notebook for the function to match the format I am using for the other loss functions in this issue, so that we can standardise this procedure, which will be helpful later on when adding the PR templates.

**Comment by kartikdutt18** [comment]
The cosine embedding that I implemented ended up being different from the PyTorch implementation. The PyTorch implementation takes in three inputs however after some discussion it was decided that we don't need to do that. The third input corresponded to 1 or -1 (i.e. similarity or dissimilarity) so I took this as a parameter and implemented one of similarity or dissimilarity. Hence the third input was removed and not combined with any other input. Hence we will only have one gradient. 
To put it simply the third parameter did the following : 
If arr[i] == 1 -> cosine similarity
else cosine dissimilarity.
And I removed that parameter from forward and implemented it as a boolean so calculate similarity on all or dissimilarity on all batches. So only one gradient. Even if this wasn't the case the third tensor only switched functions.
Now we only have a boolean to do that. Let me know if this was clear. About the notebook, I'll try to do it this weekend.

**Comment by iamshnoo** [comment]
> Let me know if this was clear

Crystal. Thanks!

> About the notebook, I'll try to do it this weekend.

Great! Really appreciate the help.

> I think there are only warnings.

Maybe you would also like to put in a quick PR during the weekend to fix this.

**Comment by kartikdutt18** [comment]
> Great! Really appreciate the help.
> I think there are only warnings.
> Maybe you would also like to put in a quick PR during the weekend to fix this.

Done. I will create a PR with the notebook by tomorrow.

**Comment by zoq** [comment]
Let's keep this open.

**Comment by Abilityguy** [comment]
Little late here but wanted to point out that the Hinge Embedding Loss actually implements the Hinge Loss in [tensorflow ](https://www.tensorflow.org/api_docs/python/tf/keras/losses/hinge). The implementation can take both labels that are 0 or -1 and labels that are 1 or -1. If labels are 0 or 1, it temporarily converts them to -1 or 1 and calculates the loss.

However, the implementation is wrong and fails for multi dimensional tensors because it does not consider column major ordering of armadillo.
